### PR TITLE
[O2B-1172] Add links to ECS environment

### PR DESCRIPTION
**Pull request by xsalonx**
#### I have a JIRA ticket
- [x] branch and/or PR name(s) include(s) JIRA ID
- [x] issue has "Fix version" assigned
- [x] issue "Status" is set to "In review"
- [x] PR labels are selected

Notable changes for users:
- Added navigation to ECS active environments f for running runs
    - from run details page
    - from run overview page
    - from environments details page
    - from environments overview page

Notable changes for developers:
- Fixed tests for QC flags creation pages

Changes made to the database:
- NA

### CODE REVIEW

File: lib/public/views/Runs/format/displayEnvironmentLinks.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import { isRunRunning } from '../../../utilities/formatting/isRunRunning.js';
+const { CONFIGURATION: { ALI_ECS_GUI_URL } } = window;
+import { h } from '/js/src/index.js';
+
+/**
+ * Format environment links
+ * @param {object} run the run
+ * @return {Componenet} links
+ */
+export const displayEnvironmentLinks = ({ environmentId: id, ...run }) => isRunRunning(run)
```
Comment (martinboulais) [inline lib/public/views/Runs/format/displayEnvironmentLinks.js:25]: I think that here you are going to complex: You should simply check that :

`timeTrgEnd ?? timeO2End ?? null === null && runDuration < MAX_RUN_DURATION`
You don't need the difference between run lost vs lost actually stopped.

You can create a function, like `isRunConsideredRunning` but in this file, because you don't reuse it elsewhere.

**Review by martinboulais — CHANGES_REQUESTED** [review]

File: lib/public/views/Runs/format/displayEnvironmentLinks.js (hunk 1 above)
Comment (xsalonx) [reply lib/public/views/Runs/format/displayEnvironmentLinks.js]: applied

**Review by graduta — CHANGES_REQUESTED** [review]
I think the link should also be added and be accessible from:
- `run-overview` dropdown list of links
- `env-overview` dropdown list of links
- `env-details` in a similar fashion to `run-detail`

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+import { tooltip } from '../popover/tooltip.js';
+import spinner from '../spinner.js';
+import { configurationService } from '../../../services/configurationService.js';
+import { h, iconWarning } from '/js/src/index.js';
+import { buildUrl } from '../../../utilities/fetch/buildUrl.js';
+
+/**
+ * Format ECS environment link
+ * @param {number} environmentId the environment id
```
Comment (graduta) [inline lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js:22]: ID as generated by AliECS is a string as it is composed of both letters and digits

Comment (xsalonx) [reply lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js]: fixed

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+export const getAliEcsEnvironmentLink = (environmentId) =>
+    configurationService.configuration$.getCurrent().match({
+        Success: ({ AliEcsGuiUrl }) => h(
+            'a',
+            { target: '_blank', href: buildUrl(AliEcsGuiUrl, { page: 'environment', id: environmentId }) },
+            'ECS',
+        ),
+        Loading: () => spinner({ size: 3, absolute: false }),
+        Failure: () => tooltip(iconWarning(), 'Error when fetching configuration'),
+        NotAsked: () => tooltip(iconWarning(), 'No configuration was asked for'),
```
Comment (martinboulais) [inline lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js:34]: Don't forget that these error messages are for end user, `error when fetching configuration` has probably no meaning for them. I think in this very specific case, we can simply return null if configuration is not available (in my own links I even removed the loading state to avoid bloating the page with spinners). ONLY IN THIS CASE.

**Review by martinboulais — CHANGES_REQUESTED** [review]

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js (hunk 3 above)
Comment (xsalonx) [reply lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js:34]: applied

Comment (martinboulais) [reply lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js]: It also applies to the Not asked.
If you go on a website an see a `No configuration was asked for` you will probably be very disappointed.

File: lib/config/services.js [hunk 4]
```diff
@@ -23,22 +23,22 @@ const {
...
             url: process.env?.EPN_INFOLOGGER_URL || null,
         },
     },
     qcGui: {
-        url: process.env?.QC_GUI_URL || changeOriginPort(origin, 8082),
+        url: process.env?.QC_GUI_URL || null,
     },
     aliFlpIndex: {
-        url: process.env?.ALI_FLP_INDEX_URL || changeOriginPort(origin, 80),
+        url: process.env?.ALI_FLP_INDEX_URL || null,
```
Comment (martinboulais) [inline lib/config/services.js:41]: Prefer using ?? instead of ||

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js [hunk 5]
```diff
@@ -12,6 +12,7 @@
  */
 
 import { configurationService } from '../../../services/configurationService.js';
+window.configurationService = configurationService
```
Comment (martinboulais) [inline lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js:15]: Probably a leftover?

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+import { configurationService } from '../../../services/configurationService.js';
+window.configurationService = configurationService
+import { h } from '/js/src/index.js';
+import { buildUrl } from '../../../utilities/fetch/buildUrl.js';
+
+/**
+ * Format ECS environment link
+ * @param {string} environmentId the environment id
+ * @return {Componenet} link
```
Comment (martinboulais) [inline lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js:22]: typo

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ * or submit itself to any jurisdiction.
+ */
+
+import { configurationService } from '../../../services/configurationService.js';
+window.configurationService = configurationService
+import { h } from '/js/src/index.js';
+import { buildUrl } from '../../../utilities/fetch/buildUrl.js';
+
+/**
+ * Format ECS environment link
```
Comment (martinboulais) [inline lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js:20]: What does that mean?

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ * @param {string} environmentId the environment id
+ * @return {Componenet} link
+ */
+export const getAliEcsEnvironmentLink = (environmentId) =>
```
Comment (martinboulais) [inline lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js:24]: Please try to have similar names than other links components (in that same directory)

File: lib/public/views/Environments/ActiveColumns/environmentsActiveColumns.js [hunk 9]
```diff
@@ -32,13 +33,16 @@ export const environmentsActiveColumns = {
...
+        format: (environmentId, { status }) => buttonLinkWithDropdown(
             environmentId,
             'env-details',
             { environmentId },
             [
                 h(CopyToClipboardComponent, { value: environmentId, id: environmentId }, 'Copy Environment Id'),
                 ...infologgerLinksComponents({ environmentId }),
+                status === 'RUNNING'
+                    ? getAliEcsEnvironmentLink(environmentId)
+                    : null,
```
Comment (martinboulais) [inline lib/public/views/Environments/ActiveColumns/environmentsActiveColumns.js:45]: I have put the logic of either the link should exist or not in QCG url provider function, not in the component, please do the same.
If you think this is a bad idea, feel free to explain

File: lib/public/views/Runs/Details/runDetailsComponent.js [hunk 10]
```diff
@@ -105,6 +109,7 @@ export const runDetailsComponent = (runDetailsModel, router) => runDetailsModel.
                     epnInfologgerLinkUrl && h('a', { href: epnInfologgerLinkUrl }, 'EPN'),
                     h('', '-'),
                     qcGuiLinkComponent(run),
+                    isRunConsideredRunning(run) && aliEcsEnvironmentLink && [h('', '-'), aliEcsEnvironmentLink],
```
Comment (martinboulais) [inline lib/public/views/Runs/Details/runDetailsComponent.js:112]: I think that we should try to be consistent, and have separated things for
- getting URL
- link component
Here for example, the links do not open links in a new tab (i don't know if it's a good thing or not).
If you think using the component is the best thing here, please update the infologger URLS to be consistents.

File: test/public/envs/detailsPage.test.js [hunk 11]
```diff
@@ -92,4 +101,19 @@ module.exports = () => {
         expect(await table[1].evaluate((row) => row.id)).to.equal('row3');
         expect(await table[2].evaluate((row) => row.id)).to.equal('row4');
     });
+
+    it('should successfully display FLP nad ECS links', async () => {
+        await goToPage(page, 'env-details', { queryParameters: { environmentId: 'CmCvjNbg' } });
```
Comment (martinboulais) [inline test/public/envs/detailsPage.test.js:106]: No more goToPage except for new test suite. It looks like pptr fails at some point if we navigate too much. If you really need to navigate from the previous page you can use click on the main links (best is to try not to navigate at all)

File: test/public/envs/overview.test.js [hunk 12]
```diff
@@ -177,4 +178,33 @@ module.exports = () => {
         await waitForNavigation(page, () => pressElement(page, '#rowTDI59So3d-runs a'));
         expectUrlParams(page, { page: 'run-detail', runNumber: 103 });
     });
+
+    it('should successfully display dropdown links', async () => {
+        await goToPage(page, 'env-overview');
```
Comment (martinboulais) [inline test/public/envs/overview.test.js:183]: No more goToPage except for new test suite.

File: test/public/runs/detail.test.js [hunk 13]
```diff
@@ -464,4 +466,26 @@ module.exports = () => {
         expect(await page.$eval('.external-links a:nth-of-type(2)', ({ href }) => href))
             .to.equal('http://localhost:8082/?page=layoutShow&runNumber=108&definition=COMMISSIONING&pdpBeamType=cosmic&runType=PHYSICS');
     });
+
+    it('should display links to environment in ECS if run is running', async () => {
+        // Test for not running run
+        await goToPage(page, 'run-detail', { queryParameters: { runNumber: 104 } });
```
Comment (martinboulais) [inline test/public/runs/detail.test.js:472]: No more goToPage except for new test suite.

File: test/public/runs/overview.test.js [hunk 14]
```diff
@@ -1099,19 +1101,33 @@ module.exports = () => {
         expect(popoverContent).to.equal('Duration based on o2 start AND stop because of missing trigger information');
     });
 
-    it('should successfully display links to infologger and QC GUI', async () => {
+    it('should successfully display links to infologger, QC GUI and ECS', async () => {
+        await runService.create({ runNumber: 1000, timeTrgStart: new Date(), environmentId: 'CmCvjNbg' });
+        await goToPage(page, 'run-overview');
```
Comment (martinboulais) [inline test/public/runs/overview.test.js:1106]: No more goToPage except for new test suite.

File: test/public/defaults.js [hunk 15]
```diff
@@ -611,6 +611,35 @@ module.exports.expectColumnValues = async (page, columnId, expectedInnerTextValu
...
+/**
+ * Method to check cells of a row with given id have expected innerText
+ *
+ * @param {puppeteer.Page} page the puppeteer page
+ * @param {stirng} rowId row id
+ * @param {Object<string, string>} [expectedInnerTextValues] values expected in the row
+ *
+ * @return {Promise<void>} resolve once row's values were checked
+ */
+module.exports.expectRowValues = async (page, rowId, expectedInnerTextValues) => {
```
Comment (martinboulais) [inline test/public/defaults.js:623]: Do not belong to this PR

File: test/public/envs/overview.test.js [hunk 16]
```diff
@@ -177,4 +178,33 @@ module.exports = () => {
...
+        // Not running env
+        envId = 'EIDO13i3D';
+        await pressElement(page, `tr[id='row${envId}'] .popover-trigger`);
+        popover = await page.waitForNavigation(`.popover:has(button[id='copy-${envId}'])`);
+        await expectLink(popover, 'a:nth-of-type(1)', {
+            href: 'http://localhost:8081/?q={%22partition%22:{%22match%22:%22EIDO13i3D%22},%22severity%22:{%22in%22:%22W%20E%20F%22}}',
+            innerText: 'Infologger FLP',
+        });
+
+        expect(await popover.$('a:nth-of-type(2)')).to.be.null;
```
Comment (martinboulais) [inline test/public/envs/overview.test.js:208]: I think it's better to use `waitForSelector('', {hidden: true});` it seems a bit more robust to me, open to debate.

File: test/public/runs/detail.test.js [hunk 17]
```diff
@@ -464,4 +466,26 @@ module.exports = () => {
...
+        await page.waitForSelector('.external-links a:nth-of-type(3)', { hidden: true, timeout: 250 });
+
+        // Create running run
+        await runService.create({ runNumber: 1010, timeTrgStart: new Date(), environmentId: 'CmCvjNbg' });
+        await goToPage(page, 'run-detail', { queryParameters: { runNumber: 1010 } });
```
Comment (martinboulais) [inline test/public/runs/detail.test.js:477]: No more goToPage except for new test suite. 
Here I see no point in not creating the run at the start of the test, or even better at the start of the test suite.
(I saw pupeteer issues when calling services directly in between tests, no idea why)

File: test/public/runs/overview.test.js [hunk 18]
```diff
@@ -1099,19 +1101,33 @@ module.exports = () => {
         expect(popoverContent).to.equal('Duration based on o2 start AND stop because of missing trigger information');
     });
 
-    it('should successfully display links to infologger and QC GUI', async () => {
+    it('should successfully display links to infologger, QC GUI and ECS', async () => {
+        await runService.create({ runNumber: 1000, timeTrgStart: new Date(), environmentId: 'CmCvjNbg' });
```
Comment (martinboulais) [inline test/public/runs/overview.test.js:1105]: Same thing, I think this would be better at the start of the test suite

**Review by martinboulais — CHANGES_REQUESTED** [review]

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js (hunk 5 above)
Comment (xsalonx) [reply lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js]: removed

File: lib/config/services.js (hunk 4 above)
Comment (xsalonx) [reply lib/config/services.js:41]: in another PR

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js (hunk 6 above)
Comment (xsalonx) [reply lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js:21]: fixed

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js (hunk 7 above)
Comment (xsalonx) [reply lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js]: like format in Active columns but changed

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js (hunk 8 above)
Comment (xsalonx) [reply lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js:23]: applied

File: lib/public/views/Environments/ActiveColumns/environmentsActiveColumns.js (hunk 9 above)
Comment (xsalonx) [reply lib/public/views/Environments/ActiveColumns/environmentsActiveColumns.js:45]: applied

File: lib/public/views/Runs/Details/runDetailsComponent.js (hunk 10 above)
Comment (xsalonx) [reply lib/public/views/Runs/Details/runDetailsComponent.js:112]: all links have `target: '_blank'` so they open it

Comment (xsalonx) [reply lib/public/views/Runs/Details/runDetailsComponent.js:112]: refactored

File: test/public/envs/detailsPage.test.js (hunk 11 above)
Comment (xsalonx) [reply test/public/envs/detailsPage.test.js:106]: refactored

File: test/public/envs/overview.test.js (hunk 12 above)
Comment (xsalonx) [reply test/public/envs/overview.test.js]: applied

File: test/public/runs/detail.test.js (hunk 13 above)
Comment (xsalonx) [reply test/public/runs/detail.test.js]: applied

File: test/public/runs/detail.test.js (hunk 17 above)
Comment (xsalonx) [reply test/public/runs/detail.test.js]: refactored

File: test/public/runs/overview.test.js (hunk 18 above)
Comment (xsalonx) [reply test/public/runs/overview.test.js:1107]: applied

File: lib/public/views/Environments/ActiveColumns/environmentsActiveColumns.js (hunk 9 above)
Comment (martinboulais) [reply lib/public/views/Environments/ActiveColumns/environmentsActiveColumns.js]: It do not looks like it?

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js (hunk 6 above)
Comment (martinboulais) [reply lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js]: I still see Componenet

File: lib/public/views/Environments/Details/EnvironmentDetailsPage.js [hunk 19]
```diff
@@ -75,10 +78,9 @@ export const EnvironmentDetailsPage = ({ envs: { detailsModel } }) => detailsMod
...
                 h('#environment-creation-date', [h('strong', 'Created at '), formatTimestamp(createdAt)]),
             ]),
-            (flpInfologgerLinkUrl || epnInfologgerLinkUrl) && h('h3.flex-row.w-100.g2.items-baseline.mb3', [
-                h('', 'Infologger: '),
-                flpInfologgerLinkUrl && h('a', { href: flpInfologgerLinkUrl }, 'FLP'),
-                epnInfologgerLinkUrl && h('a', { href: epnInfologgerLinkUrl }, 'EPN'),
+            h('h3.flex-row.w-100.g2.items-baseline.mb3', [
+                (flpInfologgerLink || epnInfologgerLink) && [h('', 'Infologger: '), flpInfologgerLink, epnInfologgerLink],
+                status === 'RUNNING' && aliEcsEnvironmentLink && [h('', '-'), aliEcsEnvironmentLink],
             ]),
```
Comment (martinboulais) [inline lib/public/views/Environments/Details/EnvironmentDetailsPage.js:84]: Here you probably need a flatMap (like https://github.com/AliceO2Group/Bookkeeping/blob/f1432d6fb2bbfcc5604d50f0e4da30a50e65c872/lib/public/components/environments/environmentStatusHistoryComponent.js#L39) because if there is no infologger links, you will have an issue

**Review by martinboulais — DISMISSED** [review]

File: lib/public/views/Environments/Details/EnvironmentDetailsPage.js (hunk 19 above)
Comment (xsalonx) [reply lib/public/views/Environments/Details/EnvironmentDetailsPage.js]: applied

File: test/public/envs/overview.test.js (hunk 16 above)
Comment (xsalonx) [reply test/public/envs/overview.test.js:204]: applied

File: lib/public/views/Environments/Details/EnvironmentDetailsPage.js (hunk 19 above)
Comment (martinboulais) [reply lib/public/views/Environments/Details/EnvironmentDetailsPage.js]: No. Please carefully look at how it is implemented in the example I sent.

File: lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js (hunk 6 above)
Comment (xsalonx) [reply lib/public/components/common/externalLinks/getAliEcsEnvironmentLink.js]: now, fixed

File: lib/public/views/Environments/ActiveColumns/environmentsActiveColumns.js (hunk 9 above)
Comment (xsalonx) [reply lib/public/views/Environments/ActiveColumns/environmentsActiveColumns.js]: ok, I got what you meant, no, it won't be as there are different decision criteria for environments and runs also runs don't include environment entity so cannot do it one way

File: lib/public/views/Environments/Details/EnvironmentDetailsPage.js (hunk 19 above)
Comment (xsalonx) [reply lib/public/views/Environments/Details/EnvironmentDetailsPage.js]: It because I don't think exactly the same approach is better, but ok applied

**Review by martinboulais — APPROVED** [review]
